packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts

```diff
@@ -0,0 +1,621 @@
+  const isEmptyString = isString && value.trim() === '';
+  return isString && !isEmptyString;
+}
+
+/**
+ * Invokes a function, passing an instance of the `LCPImageObserver` as an argument.
+ *
+ * Notes:
+ * - the `LCPImageObserver` is a tree-shakable provider, provided in 'root',
+ *   thus it's a singleton within this application
+ * - the process of `LCPImageObserver` creation and an actual operation are invoked outside of the
+ *   NgZone to make sure none of the calls inside the `LCPImageObserver` class trigger unnecessary
+ *   change detection
+ */
+function withLCPImageObserver(
+    injector: Injector, operation: (observer: LCPImageObserver) => void): void {
+  const ngZone = injector.get(NgZone);
+  return ngZone.runOutsideAngular(() => {
+    const observer = injector.get(LCPImageObserver);
+    operation(observer);
+  });
+}
+
+/***** Assert functions *****/
+
+// Verifies that there is no `src` set on a host element.
+function assertNoConflictingSrc(dir: NgOptimizedImage) {
+  if (dir.src) {
+    throw new RuntimeError(
+        RuntimeErrorCode.UNEXPECTED_SRC_ATTR,
+        `${imgDirectiveDetails(dir.rawSrc)} both \`src\` and \`rawSrc\` have been set. ` +
+            `Supplying both of these attributes is not necessary and will break lazy loading. ` +
+            `The NgOptimizedImage directive will set \`src\` itself based on the value of \`rawSrc\`. ` +
+            `To fix this, please remove the \`src\` attribute.`);
+  }
+}
+
+// Verifies that there is no `srcset` set on a host element.
+function assertNoConflictingSrcset(dir: NgOptimizedImage) {
+  if (dir.srcset) {
+    throw new RuntimeError(
+        RuntimeErrorCode.UNEXPECTED_SRCSET_ATTR,
+        `${imgDirectiveDetails(dir.rawSrc)} both \`srcset\` and \`rawSrcset\` have been set. ` +
+            `Supplying both of these attributes is not necessary and will break lazy loading. ` +
+            `The NgOptimizedImage directive will set \`srcset\` itself based on the value of ` +
+            `\`rawSrcset\`. To fix this, please remove the \`srcset\` attribute.`);
+  }
+}
+
+// Verifies that the `rawSrc` is not a Base64-encoded image.
+function assertNotBase64Image(dir: NgOptimizedImage) {
+  let rawSrc = dir.rawSrc.trim();
+  if (rawSrc.startsWith('data:')) {
+    if (rawSrc.length > BASE64_IMG_MAX_LENGTH_IN_ERROR) {
+      rawSrc = rawSrc.substring(0, BASE64_IMG_MAX_LENGTH_IN_ERROR) + '...';
+    }
+    throw new RuntimeError(
+        RuntimeErrorCode.INVALID_INPUT,
+        `${imgDirectiveDetails(dir.rawSrc, false)} \`rawSrc\` is a Base64-encoded string ` +
+            `(${rawSrc}). Base64-encoded strings are not supported by the NgOptimizedImage directive. ` +
+            `To fix this, disable the NgOptimizedImage directive for this element ` +
+            `by removing \`rawSrc\` and using a regular \`src\` attribute instead.`);
+  }
+}
+
+// Verifies that the `rawSrc` is not a Blob URL.
+function assertNotBlobURL(dir: NgOptimizedImage) {
```

Comment:
```suggestion
function assertNotBlobUrl(dir: NgOptimizedImage) {
```
(and any similar functions elsewhere)